### DISCUSSION TITLE
Hosting on AWS S3/CloudFront

### DISCUSSION THREAD

**Discussion by TheBrambleShark** [statiqdev/Discussions#82] [Q&A]
This is an issue with a few potential solutions. I thought I'd ask here though, just in case anyone here has any ideas.

I am hosting two versions of my site:
https://dev.luzfaltex.com - GitHub Pages
https://www.luzfaltex.com - S3 + CloudFront

On the GitHub Pages version, the site works perfectly. This is also seen when using the `preview` command line argument. When using the AWS site any pages in subfolders are inaccessible (404).

In order to replicate URL rewriting, I am using this lambda@edge script:

```js
'use strict';

const pointsToFile = uri => /\/[^/]+\.[^/]+$/.test(uri);
const hasTrailingSlash = uri => uri.endsWith('/');

exports.handler = (event, context, callback) => {
    // Extract the request from the CloudFront event that is sent to Lambda@Edge
    var request = event.Records[0].cf.request;

    // Extract the URI and query string from the request
    const olduri = request.uri;
    const qs = request.querystring;

    if (pointsToFile(olduri)) {
        callback(null, request);
        return;
    }

    // Append ".html" extension
    if (!hasTrailingSlash(olduri)) {
        request.uri = olduri + ".html";
    } else {
        // Append "index.html"
        request.uri = olduri + "index.html"
    }

    // Return to CloudFront
    return callback(null, request);
}
```

In my [_MasterLayout.cshtml](https://github.com/LuzFaltex/luzfaltex.com/blob/main/input/_MasterLayout.cshtml#L39-L93) file, I have links being built like so:

```cshtml
@{
    // Get the links
    string indexUrl = ExecutionContext.GetLink("index.html");
    string projectsUrl = ExecutionContext.GetLink("projects.html");
    string docsUrl = ExecutionContext.GetLink("docs/index.html");
    string policiesUrl = ExecutionContext.GetLink("policies/index.html");
    string contactUrl = ExecutionContext.GetLink("contact/index.html");
    string aboutUsUrl = ExecutionContext.GetLink("about");
    string privacyPolicyUrl = ExecutionContext.GetLink("policies/privacy.html") + "#cookies";

    // Determine if the navbar button is active
    var docPath = Document.GetLink();
    string indexActive = (docPath == indexUrl) ? "is-active" : "not-active";
    string projectsActive = docPath == projectsUrl ? "is-active" : "not-active";
    string docsActive = (docPath.StartsWith(docsUrl) || docPath.StartsWith(policiesUrl)) ? "is-active" : "not-active";
    string contactActive = docPath.StartsWith(contactUrl) ? "is-active" : "not-active";
    string aboutUsActive = docPath == aboutUsUrl ? "is-active" : "not-active";
}

<!-- snip -->
<!-- Build the navbar links -->
<ul>                                
    <li><a class="@indexActive" href="@indexUrl">Home</a></li>
    <li><a class="@projectsActive" href="@projectsUrl">Projects</a></li>
    <li><a class="@docsActive" href="@docsUrl">Docs</a></li>
    <li><a class="@aboutUsActive" href="@aboutUsUrl">About Us</a></li>
    <li><a class="@contactActive" href="@contactUrl">Contact</a></li>
</ul>
```

When the navbar links are built, it looks like this (sample from index.html):

```html
<ul>                                
    <li><a class="is-active" href="/">Home</a></li>
    <li><a class="not-active" href="/projects">Projects</a></li>
    <li><a class="not-active" href="/docs">Docs</a></li>
    <li><a class="not-active" href="/about">About Us</a></li>
    <li><a class="not-active" href="/contact">Contact</a></li>
</ul>
```

On pages like `projects.html` or `about.html`, the rewrite rule works perfectly. It sees that you're going to `/projects`, the url does not end with a forward slash, and so it appends `.html`. The url shows `/projects` but you're actually being served `/projects.html`.

When going to the docs page on the other hand, it sees the same thing and attempts to server `/docs.html`, which is not a file that exists. The actual file it should serve here is `/docs/index.html`.

On the preview web server and on GitHub Pages, this is handled automatically and the correct page is served, even though it is the same HTML.

I see a few solutions to this:

1. Manually append a forward slash in navbar links. This will cause issues down the road with it being tempermental as far as links are concerned.
2. Replace the url rewrite method with one that "just figures it out." Unfortunately, I can't find much in the way of documentation here for how I might do that.
3. Get Statiq to somehow export links as `/docs/` when it's going to an index file in a subdirectory.
4. Replace the url rewrite method to always append "index.html" if no explicit file is mentioned, configure Statiq to always export links as fully qualified, and deal with the ugly urls.

If anyone has any other ideas of how I may get around this issue and have the site behave as it does on GitHub Pages, I'm all ears.

(For bonus points, here are is the .htaccess rewrite rule I would be using)
```.htaccess
# Don't show file names
RewriteCond %{SCRIPT_FILENAME} !-d
RewriteCond %{SCRIPT_FILENAME} !-f
RewriteRule ^(.*)$ index.html?$1 [L,QSA]
```

Edit: I have attempted to rewrite the url rewrite script. This is the current version:

```js
'use strict';

const pointsToFile = uri => /\/[^/]+\.[^/]+$/.test(uri);
const hasTrailingSlash = uri => uri.endsWith('/');

var AWS = require('aws-sdk');
var S3 = new AWS.S3();

exports.handler = (event, context, callback) => {
    // Extract the request from the CloudFront event that is sent to Lambda@Edge 
    var request = event.Records[0].cf.request;

    // Extract the URI and query string from the request
    const olduri = request.uri;

    if (pointsToFile(olduri)) {
        // Specific file requested. Do nothing.
        return callback(null, request);
    }

    if (hasTrailingSlash(olduri)) {
        // If it has a trailing slash, then we know
        // it's a directory and we should look for index.
        request.uri = olduri + "index.html";
        return callback(null, request);
    } 
    
    // Get the bucket name
    const src_bkt = event.Records[0].s3.bucket.name;

    // Let's try finding the file simply by appending .html
    const key = olduri + ".html";

    S3.headObject({ src_bkt, key }, function(err, data) {
        if (err) {
            // Swallow the error.
            // File does not exist or S3 is down.
            // Assume the requested endpoint is a directory
            // and append /index.html.
            request.uri = olduri + "/index.html";
        } else {
            // File was found, so just map to that file.
            request.uri = key;
        }
    });

    // Return to CloudFront
    return callback(null, request);
};
```

**Comment by daveaglick** [comment]
This is really interesting, thanks for all the detail. Once we figure it out, I'd like to put what you end up with working on the Statiq site if that's okay.

I actually don't know anything about AWS hosting myself - I generally use a mix of GitHub Pages, Azure App Service, and Netlify. What I do know is that unlike GitHub Pages and Netlify, which are designed to support static sites, Azure App Service also requires some rewrite rules (at the bottom of https://statiq.dev/web/deployment/azure-app-service). As you've found, there are basically two separate conventions that need to be supported:

- Extensionless URLs (I.e. appending .html to the name of a file, which your rewrite rule is handling)
- Default files (I.e. `index.html`, which it looks like AWS doesn't handle out-of-the-box - IIS/Azure already supports index files by default)

I wonder if there's some higher-level setting in AWS that turns on default index files for a site, before the rewrite rule even gets called? I.e. can you set `DirectoryIndex` in your `.htaccess` file:

```
DirectoryIndex index.html index.htm
```

**Reply by TheBrambleShark** [reply]
I thought I'd replied to this but it turns out I had not!

Unfortunately, AWS has nothing like an .htaccess file. Instead, it uses javascript-based firewall routing rules, such as the one provided in my example, to determine how it accesses files in S3.

In the end, it became too much of a headache and I ended up just changing hosting providers to NearlyFreeSpeech.net
